frontend/src/lib/utils.ts

```diff
@@ -203,3 +203,93 @@ export function labelVisibilityProtoValueToEnum(
       return LabelVisibilityOptions.Visible
   }
 }
+
+/**
+ * Looks for an IFrame with given className inside given querySet
+ */
+export function findAnIFrameWithClassName(
+  qs: NodeListOf<HTMLIFrameElement> | HTMLCollectionOf<HTMLIFrameElement>,
+  className: string
+): HTMLIFrameElement | null {
+  for (let i = 0; i < qs.length; i++) {
+    const cd = qs[i].contentDocument
+    if (cd && cd.getElementsByClassName(className).length > 0) {
+      return qs[i]
+    }
+  }
+  return null
+}
+
+/**
+ * Returns True if IFrame can be accessed otherwise returns False
+ */
+export function canAccessIFrame(iframe: HTMLIFrameElement): boolean {
+  try {
+    if (iframe.contentWindow === null) return false
+    const doc = iframe.contentDocument || iframe.contentWindow.document
+    const html = doc.body.innerHTML
+    return html !== null && html !== ""
+  } catch (err) {
+    return false
+  }
+}
+
+/**
+ * Tries to get an IFrame in which Streamlit app is embedded on Cloud deployments.
+ * It assumes iframe has title="streamlitApp", iterates over IFrames,
+ * and looks which IFrame contains div with stAppId value, otherwise returns first found iFrame or null.
+ */
+export function getIFrameEnclosingApp(
+  embeddingId: string
+): HTMLIFrameElement | null {
+  if (!isInChildFrame()) {
+    return null
+  }
+  const embeddingIdClassName = getEmbeddingIdClassName(embeddingId)
+  const qsStreamlitAppStr = 'iframe[title="streamlitApp"]'
+  let qs = window.document.querySelectorAll(
+    qsStreamlitAppStr
+  ) as NodeListOf<HTMLIFrameElement>
+  if (qs.length < 1 && window.parent) {
+    qs = window.parent.document.querySelectorAll(qsStreamlitAppStr)
```

Comment:
Will a browser allow this line to be called from an iframe from a different origin?

I'm not very familiar with exactly where cross-origin restrictions would kick in, but I'm a bit surprised this would be possible since we're accessing things in `window.parent`

Reply:
Just to be `100%` sure I've updated `printCallback` to wrap all of controversial calls inside `try catch` 👍